forecasting/src/autogluon/forecasting/dataset/ts_dataframe.py

```diff
@@ -0,0 +1,331 @@
+    """
+
+    def __init__(self, data: Any, *args, **kwargs):
+        if isinstance(data, TimeSeriesListDataset):
+            data = self.from_list_dataset(data)
+        if isinstance(data, pd.DataFrame):
+            if not isinstance(data.index, pd.MultiIndex):
+                data = self.from_data_frame(data)
+        self._validate_multi_index_data_frame(data)
+        super().__init__(data=data, *args, **kwargs)
+
+    @classmethod
+    def example(cls):
+        """An example TimeSeriesDataFrame.
+
+        Returns
+        -------
+        ts_df : TimeSeriesDataFrame
+            It returns an example TimeSeriesDataFrame as:
+                                    target
+            item_id timestamp
+            0       2019-01-01       0
+                    2019-01-02       1
+                    2019-01-03       2
+            1       2019-01-01       3
+                    2019-01-02       4
+                    2019-01-03       5
+            2       2019-01-01       6
+                    2019-01-02       7
+                    2019-01-03       8
+        """
+
+        target = np.arange(9)
+        datetime_index = tuple(
+            pd.date_range(pd.Timestamp("01-01-2019"), periods=3, freq="D")
+        )
+        item_ids = (0, 1, 2)
+        multi_index = pd.MultiIndex.from_product(
+            [item_ids, datetime_index], names=["item_id", "timestamp"]
+        )
+        return TimeSeriesDataFrame(
+            pd.Series(target, name="target", index=multi_index).to_frame()
+        )
+
+    @classmethod
+    def _validate_data_frame(cls, df: pd.DataFrame):
+        if not isinstance(df, pd.DataFrame):
+            raise ValueError(f"data must be a pd.DataFrame, got {type(df)}")
+        if "item_id" not in df.columns:
+            raise ValueError(f"data must have a `item_id` column")
+        if "timestamp" not in df.columns:
+            raise ValueError(f"data must have a `timestamp` column")
+        if df["item_id"].isnull().any():
+            raise ValueError(f"`item_id` columns can not have nan")
+        if df["timestamp"].isnull().any():
+            raise ValueError(f"`timestamp` columns can not have nan")
+        if not df["item_id"].dtype == "int64":
+            raise ValueError(f"for item_id, the only pandas dtype allowed is ‘int64’.")
+        if not df["timestamp"].dtype == "datetime64[ns]":
+            raise ValueError(
+                f"for timestamp, the only pandas dtype allowed is ‘datetime64[ns]’."
+            )
+
+    @classmethod
+    def _validate_list_dataset(cls, list_dataset):
+        if not isinstance(list_dataset, TimeSeriesListDataset):
+            raise ValueError(
+                f"list_dataset must be a TimeSeriesListDataset, got {type(list_dataset)}"
+            )
+        if len(list_dataset.data) == 0:
+            raise ValueError(f"list_dataset has no time-series.")
+        for i, ts in enumerate(list_dataset.data):
```

Comment:
as we are now using `UserList`, all references to `list_dataset.data` could be replaced by `list_dataset`. Similarly for `len` checks above and other iterators below.